File: bindings/kotlin/src/test/kotlin/zag/ZagDiscoverTests.kt
package zag

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue
import kotlinx.coroutines.runBlocking

/**
 * Integration tests for ZagDiscover.
 *
 * These tests require the zag binary to be built and available in PATH.
 */
class ZagDiscoverTests {

    @Test
    fun `listProviders returns provider names`() = runBlocking {
        val providers = ZagDiscover.listProviders()
        assertTrue(providers.size >= 5)
        assertTrue(providers.contains("claude"))
        assertTrue(providers.contains("codex"))
        assertTrue(providers.contains("gemini"))
        assertTrue(providers.contains("copilot"))
        assertTrue(providers.contains("ollama"))
    }

    @Test
    fun `getCapability returns single provider`() = runBlocking {
        val cap = ZagDiscover.getCapability("claude")
        assertEquals("claude", cap.provider)
        assertTrue(cap.availableModels.isNotEmpty())
        assertTrue(cap.features.interactive.supported)
    }

    @Test
    fun `getAllCapabilities returns all providers`() = runBlocking {
        val caps = ZagDiscover.getAllCapabilities()
        assertTrue(caps.size >= 5)
        val names = caps.map { it.provider }
        assertTrue(names.contains("claude"))
    }

    @Test
    fun `resolveModel resolves alias`() = runBlocking {
        val rm = ZagDiscover.resolveModel("claude", "small")
        assertEquals("small", rm.input)
        assertEquals("haiku", rm.resolved)
        assertTrue(rm.isAlias)
        assertEquals("claude", rm.provider)
    }

    @Test
    fun `resolveModel passes through non-alias`() = runBlocking {
        val rm = ZagDiscover.resolveModel("claude", "opus")
        assertEquals("opus", rm.input)
        assertEquals("opus", rm.resolved)
        assertFalse(rm.isAlias)
    }
}

/**
 * Unit tests for discover model deserialization.
 */
class DiscoverModelsTests {

    @Test
    fun `FeatureSupport deserializes`() {
        val json = """{"supported":true,"native":false}"""
        val fs = ZagJson.decodeFromString<FeatureSupport>(json)
        assertTrue(fs.supported)
        assertFalse(fs.isNative)
    }

    @Test
    fun `SessionLogSupport deserializes`() {
        val json = """{"supported":true,"native":true,"completeness":"full"}"""
        val sls = ZagJson.decodeFromString<SessionLogSupport>(json)
        assertTrue(sls.supported)
        assertTrue(sls.isNative)
        assertEquals("full", sls.completeness)
    }

    @Test
    fun `SessionLogSupport with null completeness`() {
        val json = """{"supported":false,"native":false}"""
        val sls = ZagJson.decodeFromString<SessionLogSupport>(json)
        assertFalse(sls.supported)
        assertEquals(null, sls.completeness)
    }

    @Test
    fun `StreamingInputSupport deserializes with semantics`() {
        val json = """{"supported":true,"native":true,"semantics":"queue"}"""
        val sis = ZagJson.decodeFromString<StreamingInputSupport>(json)
        assertTrue(sis.supported)
        assertTrue(sis.isNative)
        assertEquals("queue", sis.semantics)
    }

    @Test
    fun `StreamingInputSupport with null semantics`() {
        val json = """{"supported":false,"native":false}"""
        val sis = ZagJson.decodeFromString<StreamingInputSupport>(json)
        assertFalse(sis.supported)
        assertEquals(null, sis.semantics)
    }

    @Test
    fun `SizeMappings deserializes`() {
        val json = """{"small":"haiku","medium":"sonnet","large":"opus"}"""
        val sm = ZagJson.decodeFromString<SizeMappings>(json)
        assertEquals("haiku", sm.small)
        assertEquals("sonnet", sm.medium)
        assertEquals("opus", sm.large)
    }

    @Test
    fun `ProviderCapability deserializes`() {
        val json = """
        {
            "provider": "claude",
            "default_model": "sonnet",
            "available_models": ["haiku", "sonnet", "opus"],
            "size_mappings": {"small": "haiku", "medium": "sonnet", "large": "opus"},
            "features": {
                "interactive": {"supported": true, "native": true},
                "non_interactive": {"supported": true, "native": true},
                "resume": {"supported": true, "native": true},
                "resume_with_prompt": {"supported": true, "native": true},
                "session_logs": {"supported": true, "native": true, "completeness": "full"},
                "json_output": {"supported": true, "native": true},
                "stream_json": {"supported": true, "native": true},
                "json_schema": {"supported": false, "native": false},
                "input_format": {"supported": true, "native": true},
                "streaming_input": {"supported": true, "native": true, "semantics": "queue"},
                "worktree": {"supported": true, "native": false},
                "sandbox": {"supported": true, "native": false},
                "system_prompt": {"supported": true, "native": true},
                "auto_approve": {"supported": true, "native": true},
                "review": {"supported": true, "native": true},
                "add_dirs": {"supported": true, "native": true},
                "max_turns": {"supported": true, "native": true}
            }
        }
        """.trimIndent()

        val cap = ZagJson.decodeFromString<ProviderCapability>(json)
        assertEquals("claude", cap.provider)
        assertEquals("sonnet", cap.defaultModel)
        assertEquals(3, cap.availableModels.size)
        assertEquals("haiku", cap.sizeMappings.small)
        assertTrue(cap.features.interactive.supported)
        assertTrue(cap.features.interactive.isNative)
        assertTrue(cap.features.sessionLogs.supported)
        assertEquals("full", cap.features.sessionLogs.completeness)
        assertFalse(cap.features.jsonSchema.supported)
        assertTrue(cap.features.worktree.supported)
        assertFalse(cap.features.worktree.isNative)
        assertTrue(cap.features.streamingInput.supported)
        assertEquals("queue", cap.features.streamingInput.semantics)
    }

    @Test
    fun `ResolvedModel deserializes`() {
        val json = """{"input":"small","resolved":"haiku","is_alias":true,"provider":"claude"}"""
        val rm = ZagJson.decodeFromString<ResolvedModel>(json)
        assertEquals("small", rm.input)
        assertEquals("haiku", rm.resolved)
        assertTrue(rm.isAlias)
        assertEquals("claude", rm.provider)
    }

    @Test
    fun `ResolvedModel non-alias`() {
        val json = """{"input":"opus","resolved":"opus","is_alias":false,"provider":"claude"}"""
        val rm = ZagJson.decodeFromString<ResolvedModel>(json)
        assertEquals("opus", rm.input)
        assertEquals("opus", rm.resolved)
        assertFalse(rm.isAlias)
    }
}

File: bindings/swift/Tests/ZagTests/ZagDiscoverTests.swift
import Foundation
import Testing
@testable import Zag

// MARK: - Discover model deserialization tests

@Suite("DiscoverModels")
struct DiscoverModelsTests {

    static let capabilityJSON = """
    {
        "provider": "claude",
        "default_model": "sonnet",
        "available_models": ["haiku", "sonnet", "opus"],
        "size_mappings": {
            "small": "haiku",
            "medium": "sonnet",
            "large": "opus"
        },
        "features": {
            "interactive": {"supported": true, "native": true},
            "non_interactive": {"supported": true, "native": true},
            "resume": {"supported": true, "native": true},
            "resume_with_prompt": {"supported": true, "native": true},
            "session_logs": {"supported": true, "native": true, "completeness": "full"},
            "json_output": {"supported": true, "native": true},
            "stream_json": {"supported": true, "native": true},
            "json_schema": {"supported": true, "native": true},
            "input_format": {"supported": true, "native": true},
            "streaming_input": {"supported": true, "native": true, "semantics": "queue"},
            "worktree": {"supported": true, "native": false},
            "sandbox": {"supported": true, "native": false},
            "system_prompt": {"supported": true, "native": true},
            "auto_approve": {"supported": true, "native": true},
            "review": {"supported": false, "native": false},
            "add_dirs": {"supported": true, "native": true},
            "max_turns": {"supported": true, "native": true}
        }
    }
    """

    static let allCapabilitiesJSON = """
    [
        {
            "provider": "claude",
            "default_model": "sonnet",
            "available_models": ["haiku", "sonnet", "opus"],
            "size_mappings": {"small": "haiku", "medium": "sonnet", "large": "opus"},
            "features": {
                "interactive": {"supported": true, "native": true},
                "non_interactive": {"supported": true, "native": true},
                "resume": {"supported": true, "native": true},
                "resume_with_prompt": {"supported": true, "native": true},
                "session_logs": {"supported": true, "native": true, "completeness": "full"},
                "json_output": {"supported": true, "native": true},
                "stream_json": {"supported": true, "native": true},
                "json_schema": {"supported": true, "native": true},
                "input_format": {"supported": true, "native": true},
                "streaming_input": {"supported": true, "native": true, "semantics": "queue"},
                "worktree": {"supported": true, "native": false},
                "sandbox": {"supported": true, "native": false},
                "system_prompt": {"supported": true, "native": true},
                "auto_approve": {"supported": true, "native": true},
                "review": {"supported": false, "native": false},
                "add_dirs": {"supported": true, "native": true},
                "max_turns": {"supported": true, "native": true}
            }
        },
        {
            "provider": "codex",
            "default_model": "codex-mini",
            "available_models": ["codex-mini"],
            "size_mappings": {"small": "codex-mini", "medium": "codex-mini", "large": "codex-mini"},
            "features": {
                "interactive": {"supported": true, "native": true},
                "non_interactive": {"supported": true, "native": true},
                "resume": {"supported": false, "native": false},
                "resume_with_prompt": {"supported": false, "native": false},
                "session_logs": {"supported": false, "native": false},
                "json_output": {"supported": true, "native": true},
                "stream_json": {"supported": false, "native": false},
                "json_schema": {"supported": false, "native": false},
                "input_format": {"supported": false, "native": false},
                "streaming_input": {"supported": false, "native": false},
                "worktree": {"supported": false, "native": false},
                "sandbox": {"supported": false, "native": false},
                "system_prompt": {"supported": false, "native": false},
                "auto_approve": {"supported": true, "native": true},
                "review": {"supported": false, "native": false},
                "add_dirs": {"supported": false, "native": false},
                "max_turns": {"supported": false, "native": false}
            }
        }
    ]
    """

    static let resolvedModelJSON = """
    {
        "input": "small",
        "resolved": "haiku",
        "is_alias": true,
        "provider": "claude"
    }
    """

    @Test("ProviderCapability deserializes from JSON")
    func providerCapabilityDeserializes() throws {
        let data = DiscoverModelsTests.capabilityJSON.data(using: .utf8)!
        let cap = try JSONDecoder().decode(ProviderCapability.self, from: data)

        #expect(cap.provider == "claude")
        #expect(cap.defaultModel == "sonnet")
        #expect(cap.availableModels == ["haiku", "sonnet", "opus"])
        #expect(cap.sizeMappings.small == "haiku")
        #expect(cap.sizeMappings.medium == "sonnet")
        #expect(cap.sizeMappings.large == "opus")
    }

    @Test("Features deserialize all 17 fields")
    func featuresDeserializeAllFields() throws {
        let data = DiscoverModelsTests.capabilityJSON.data(using: .utf8)!
        let cap = try JSONDecoder().decode(ProviderCapability.self, from: data)
        let f = cap.features

        #expect(f.interactive.supported == true)
        #expect(f.interactive.native == true)
        #expect(f.nonInteractive.supported == true)
        #expect(f.resume.supported == true)
        #expect(f.resumeWithPrompt.supported == true)
        #expect(f.sessionLogs.supported == true)
        #expect(f.sessionLogs.completeness == "full")
        #expect(f.jsonOutput.supported == true)
        #expect(f.streamJson.supported == true)
        #expect(f.jsonSchema.supported == true)
        #expect(f.inputFormat.supported == true)
        #expect(f.streamingInput.supported == true)
        #expect(f.streamingInput.semantics == "queue")
        #expect(f.worktree.supported == true)
        #expect(f.worktree.native == false)
        #expect(f.sandbox.supported == true)
        #expect(f.systemPrompt.supported == true)
        #expect(f.autoApprove.supported == true)
        #expect(f.review.supported == false)
        #expect(f.addDirs.supported == true)
        #expect(f.maxTurns.supported == true)
    }

    @Test("StreamingInputSupport without semantics")
    func streamingInputSupportWithoutSemantics() throws {
        let json = """
        {"supported": false, "native": false}
        """
        let data = json.data(using: .utf8)!
        let sis = try JSONDecoder().decode(StreamingInputSupport.self, from: data)
        #expect(sis.supported == false)
        #expect(sis.native == false)
        #expect(sis.semantics == nil)
    }

    @Test("SessionLogSupport without completeness")
    func sessionLogSupportWithoutCompleteness() throws {
        let json = """
        {"supported": true, "native": false}
        """
        let data = json.data(using: .utf8)!
        let sls = try JSONDecoder().decode(SessionLogSupport.self, from: data)
        #expect(sls.supported == true)
        #expect(sls.native == false)
        #expect(sls.completeness == nil)
    }

    @Test("array of ProviderCapability deserializes")
    func allCapabilitiesDeserialize() throws {
        let data = DiscoverModelsTests.allCapabilitiesJSON.data(using: .utf8)!
        let caps = try JSONDecoder().decode([ProviderCapability].self, from: data)

        #expect(caps.count == 2)
        #expect(caps[0].provider == "claude")
        #expect(caps[1].provider == "codex")
    }

    @Test("ResolvedModel deserializes from JSON")
    func resolvedModelDeserializes() throws {
        let data = DiscoverModelsTests.resolvedModelJSON.data(using: .utf8)!
        let rm = try JSONDecoder().decode(ResolvedModel.self, from: data)

        #expect(rm.input == "small")
        #expect(rm.resolved == "haiku")
        #expect(rm.isAlias == true)
        #expect(rm.provider == "claude")
    }

    @Test("ResolvedModel non-alias")
    func resolvedModelNonAlias() throws {
        let json = """
        {"input": "opus", "resolved": "opus", "is_alias": false, "provider": "claude"}
        """
        let data = json.data(using: .utf8)!
        let rm = try JSONDecoder().decode(ResolvedModel.self, from: data)

        #expect(rm.input == "opus")
        #expect(rm.resolved == "opus")
        #expect(rm.isAlias == false)
    }

    @Test("SizeMappings deserializes")
    func sizeMappingsDeserializes() throws {
        let json = """
        {"small": "haiku", "medium": "sonnet", "large": "opus"}
        """
        let data = json.data(using: .utf8)!
        let sm = try JSONDecoder().decode(SizeMappings.self, from: data)

        #expect(sm.small == "haiku")
        #expect(sm.medium == "sonnet")
        #expect(sm.large == "opus")
    }

    @Test("FeatureSupport deserializes")
    func featureSupportDeserializes() throws {
        let json = """
        {"supported": true, "native": false}
        """
        let data = json.data(using: .utf8)!
        let fs = try JSONDecoder().decode(FeatureSupport.self, from: data)

        #expect(fs.supported == true)
        #expect(fs.native == false)
    }
}

// MARK: - Discover function tests (require zag binary)

#if os(macOS) || os(Linux)
@Suite("ZagDiscover")
struct ZagDiscoverTests {

    @Test("listProviders returns provider names")
    func listProvidersReturnsNames() async throws {
        let providers = try await ZagDiscover.listProviders()
        #expect(providers.count >= 5)
        #expect(providers.contains("claude"))
        #expect(providers.contains("codex"))
        #expect(providers.contains("gemini"))
        #expect(providers.contains("copilot"))
        #expect(providers.contains("ollama"))
    }

    @Test("getCapability returns single provider")
    func getCapabilityReturnsSingleProvider() async throws {
        let cap = try await ZagDiscover.getCapability(provider: "claude")
        #expect(cap.provider == "claude")
        #expect(cap.availableModels.count > 0)
        #expect(cap.features.interactive.supported == true)
    }

    @Test("getAllCapabilities returns all providers")
    func getAllCapabilitiesReturnsAll() async throws {
        let caps = try await ZagDiscover.getAllCapabilities()
        #expect(caps.count >= 5)
        let names = caps.map { $0.provider }
        #expect(names.contains("claude"))
    }

    @Test("resolveModel resolves alias")
    func resolveModelResolvesAlias() async throws {
        let rm = try await ZagDiscover.resolveModel(provider: "claude", model: "small")
        #expect(rm.input == "small")
        #expect(rm.resolved == "haiku")
        #expect(rm.isAlias == true)
        #expect(rm.provider == "claude")
    }

    @Test("resolveModel passes through non-alias")
    func resolveModelPassesThrough() async throws {
        let rm = try await ZagDiscover.resolveModel(provider: "claude", model: "opus")
        #expect(rm.input == "opus")
        #expect(rm.resolved == "opus")
        #expect(rm.isAlias == false)
    }
}
#endif
